### PR TITLE
 harden configuration validation and training execution pipelines

### PR DESCRIPTION
**Pull request by AryanXCode646**
- Refactor ExperimentConfig and AlgorithmConfig with discriminated validation
- Strictly reject RL-only hyperparameters on classical planner configs
- Harden training loops in trainer.py and curriculum/trainer.py against edge cases
- Add comprehensive configuration validation test suite

### CODE REVIEW

**Review by ashishsinghbora — COMMENTED** [review]
Review — Changes Requested

I went through the configuration changes, both training pipelines, and the new validation tests. The validation improvements are useful, but I would not merge this PR as-is.

🚨 1. "AlgorithmConfig" is not actually discriminated

The PR introduces "RLAlgorithmConfig" and "PlannerAlgorithmConfig", but "ExperimentConfig" still uses the generic "AlgorithmConfig" with manual "model_validator" logic.

This leaves three models representing the same conceptual configuration.

If the goal is polymorphic/discriminated configuration, please either use the specialized models through a discriminated union or remove the unused models and keep one clearly defined approach.

🚨 2. Algorithm knowledge is becoming duplicated

"PLANNER_ALGORITHMS" now hardcodes:

{"astar", "rrt_star", "rrt*"}

while the project also has an algorithm registry.

We should have one source of truth for algorithm identity/type/capabilities. Otherwise configuration, the registry, ExperimentManager, and trainers can disagree about which algorithms are supported.

Please integrate this with the registry rather than introducing another independent list.

🚨 3. The trainers still hardcode PPO/SAC

"training/trainer.py" and "curriculum/trainer.py" still explicitly construct PPO and SAC.

That means adding a new algorithm still requires modifying the trainers even though ARL now has an algorithm registry.

Please route algorithm construction through the registry/factory so configuration → registry → runtime uses one dispatch mechanism.

🚨 4. The RRT* relational validation needs justification

This constraint:

search_radius >= 0.5 * step_size

is described as necessary for “effective near-neighbor rewiring”, but it looks like an algorithmic policy rather than a hard validity requirement.

If the underlying RRT* implementation does not mathematically/structurally require this constraint, rejecting the configuration is too restrictive for a research framework.

Please justify this against the actual implementation or remove it.

🚨 5. A* accepts a seed without establishing what it controls

"AStarParametersConfig" exposes:

seed: Optional[int]

but A* appears to be deterministic.

Since ARL is also establishing a standardized evaluation seed protocol, we should avoid exposing seed parameters that don't actually control randomness.

Please either remove this field or document and implement its actual semantics.

⚠️ 6. Validators mutate their input dictionaries

Several "mode="before"" validators modify the incoming dictionary using "pop()" and assignment.

Please normalize a copy instead:

data = dict(data)

before modifying it.

⚠️ 7. "save_config()" changes the serialization contract

Using:

exclude_none=True

means explicit "null" values disappear from saved configurations.

Given the project's reproducibility goals, please document whether omitted fields and explicit "null" are intentionally equivalent and add a regression test.

⚠️ 8. "compute_config_sha256()" needs stronger canonicalization tests

The implementation is reasonable, but tests should cover semantic equivalence cases such as:

- reordered YAML keys
- "params" vs "parameters"
- "rrt*" vs "rrt_star"
- omitted defaults vs explicit defaults
- nested dictionary ordering

Otherwise the hash contract is only partially tested.

⚠️ 9. "BaseTrainer.close()" silently swallows cleanup failures

This:

except Exception:
    pass

can hide real resource-cleanup bugs.

Please log the exception instead of silently discarding it.

⚠️ 10. Avoid relying on "assert" for runtime invariants

The new:

assert self.config.training is not None

is unnecessary after validation and can be disabled with Python optimization.

Prefer explicit invariant handling or structure the code so the type checker understands the invariant.

⚠️ 11. Add end-to-end configuration → runtime tests

The new tests do a good job checking invalid configurations, but we also need to prove that valid configurations produce the correct runtime objects.

At minimum, test:

- PPO config → PPO trainer/algorithm
- SAC config → SAC trainer/algorithm
- A* config → planner execution, not RL training
- RRT* config → planner execution with configured parameters
- unsupported algorithm → explicit failure
- planner config cannot accidentally enter an RL training path

Final verdict

REQUEST CHANGES

The validation work is useful, but the PR currently improves configuration locally while leaving algorithm dispatch duplicated across the codebase.

Before merging, I would strongly recommend making the algorithm registry the single source of truth and connecting configuration, experiment management, and training through that registry. That will prevent the framework from accumulating another layer of hardcoded algorithm knowledge as more RL algorithms and planners are added.
